File: src/main/java/ro/mpp2024/Repository/AngajatRepository.java
package ro.mpp2024.Repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.Domain.Angajat;
import ro.mpp2024.Validator.AbstractValidator;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

public class AngajatRepository implements AngajatAbstractRepository{

    private JdbcUtils dbUtils;
    private AbstractValidator<Angajat> validator;
    private static final Logger logger= LogManager.getLogger();
    public AngajatRepository(Properties props, AbstractValidator<Angajat> validator)
    {
        logger.info("Initializing AngajatRepository with properties: {} ",props);
        this.validator = validator;
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public Optional<Angajat> add(Angajat angajat) {
        logger.traceEntry("Saving angajat {}", angajat);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("insert into Angajat(nume, prenume, locatie, parola) values (?,?,?,?)"))
        {
            validator.validate(angajat);

            preparedStatement.setString(1, angajat.getNume());
            preparedStatement.setString(2, angajat.getPrenume());
            preparedStatement.setString(3, angajat.getLocatie());
            preparedStatement.setString(4, angajat.getParola());

            int rows = preparedStatement.executeUpdate();

            logger.trace("Saved {} instances", rows);
            if(rows == 0)
                return Optional.of(angajat);
            return Optional.empty();

        } catch (SQLException e) {
            logger.error(e);
            throw new RuntimeException(e);
        }
    }

    @Override
    public Optional<Angajat> update(Angajat angajat) {
        logger.traceEntry("Updating angajat {}",angajat);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("update Angajat set nume = ?, prenume = ?, locatie = ?, parola = ? where id_angajat = ?"))
        {
            validator.validate(angajat);

            preparedStatement.setString(1, angajat.getNume());
            preparedStatement.setString(2, angajat.getPrenume());
            preparedStatement.setString(3, angajat.getLocatie());
            preparedStatement.setString(4, angajat.getParola());
            preparedStatement.setInt(5, angajat.getId());

            int rows = preparedStatement.executeUpdate();

            logger.trace("Updated {} instances", rows);
            if(rows == 0)
                return Optional.of(angajat);
            return Optional.empty();

        } catch (SQLException e) {
            logger.error(e);
            throw new RuntimeException(e);
        }
    }

    @Override
    public Optional<Angajat> deleteById(Integer id) {
        logger.traceEntry("Deleting angajat with id {}",id);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("delete from Angajat where id_angajat = ?"))
        {
            Optional<Angajat> angajat = findById(id);

            preparedStatement.setInt(1, id);

            int rows = preparedStatement.executeUpdate();
            logger.trace("Deleted {} instances", rows);
            return angajat;

        } catch (SQLException e) {
            logger.error(e);
            throw new RuntimeException(e);
        }
    }

    @Override
    public Optional<Angajat> findById(Integer id) {
        logger.traceEntry("Finding angajat with id {}", id);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from Angajat where id_angajat = ?"))
        {
            preparedStatement.setInt(1, id);
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    Integer idAngajat = resultSet.getInt("id_angajat");
                    String nume = resultSet.getString("nume");
                    String prenume = resultSet.getString("prenume");
                    String locatie = resultSet.getString("locatie");
                    String parola = resultSet.getString("parola");

                    Angajat angajat = new Angajat(nume, prenume, locatie, parola);
                    angajat.setId(idAngajat);
                    logger.traceExit("Found angajat {}",angajat);
                    return Optional.of(angajat);
                }
                logger.traceExit("No angajat with id {} was found",id);
                return Optional.empty();
            }
        } catch (SQLException e) {
            logger.error(e);
            throw new RuntimeException(e);
        }
    }

    @Override
    public List<Angajat> findAll() {
        logger.traceEntry();
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from Angajat"))
        {
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                List<Angajat> angajati = new ArrayList<>();
                while(resultSet.next()){
                    Integer idAngajat = resultSet.getInt("id_angajat");
                    String nume = resultSet.getString("nume");
                    String prenume = resultSet.getString("prenume");
                    String locatie = resultSet.getString("locatie");
                    String parola = resultSet.getString("parola");

                    Angajat angajat = new Angajat(nume, prenume, locatie, parola);
                    angajat.setId(idAngajat);

                    angajati.add(angajat);
                }
                logger.traceExit("Angajati: {}", angajati);
                return angajati;
            }
        } catch (SQLException e) {
            logger.error(e);
            throw new RuntimeException(e);
        }
    }

    @Override
    public Optional<Angajat> findByFirstnameLastnamePassword(String firstname, String lastname, String password) {
        return Optional.empty();
    }
}

File: SwimmingCompetitionCSharpGUI/Repository/AngajatRepository.cs
using SwimmingCompetition.Domain;
using System.Data;
using System.Data.SqlClient;
using System.Data.SQLite;
using log4net;
using SwimmingCompetition.Validator;

namespace SwimmingCompetition.Repository;

public class AngajatRepository: IAngajatRepository
{
    private IDictionary<string, string> _props;
    private IAbstractValidator<Angajat> _validator;
    private static readonly log4net.ILog log = LogManager.GetLogger(typeof(AngajatRepository));

    public AngajatRepository(IAbstractValidator<Angajat> validator, IDictionary<string,string> props)
    {
        log.Info("Creating Angajat Repository");
        this._props = props;
        this._validator = validator;
    }

    public Angajat? Add(Angajat angajat)
    {
        log.InfoFormat("Entering Add with angajat: {0}",angajat);
        IDbConnection connection = DBUtils.getConnection(_props);
        using(var insertCommand = connection.CreateCommand())
        {
            _validator.Validate(angajat);
            
            insertCommand.CommandText = "insert into Angajat(nume, prenume, locatie, parola) values (@nume, @prenume, @locatie, @parola)";
            
            IDbDataParameter parameterNume = insertCommand.CreateParameter();
            parameterNume.ParameterName = "@nume";
            parameterNume.Value = angajat.Nume;
            insertCommand.Parameters.Add(parameterNume);
            IDbDataParameter parameterPrenume = insertCommand.CreateParameter();
            parameterPrenume.ParameterName = "@prenume";
            parameterPrenume.Value = angajat.Prenume;
            insertCommand.Parameters.Add(parameterPrenume);
            IDbDataParameter parameterLocatie = insertCommand.CreateParameter();
            parameterLocatie.ParameterName = "@locatie";
            parameterLocatie.Value = angajat.Locatie;
            insertCommand.Parameters.Add(parameterLocatie);
            IDbDataParameter parameterParola = insertCommand.CreateParameter();
            parameterParola.ParameterName = "@parola";
            parameterParola.Value = angajat.Parola;
            insertCommand.Parameters.Add(parameterParola);
            
            int rows = insertCommand.ExecuteNonQuery();
            log.InfoFormat("Inserting {0} rows", rows);
            if (rows == 0)
                return angajat;
            return null;
        }
    }

    public Angajat? Update(Angajat angajat)
    {
        log.InfoFormat("Entering Update with angajat: {0}",angajat);
        IDbConnection connection = DBUtils.getConnection(_props);
        using(var updateCommand = connection.CreateCommand())
        {
            _validator.Validate(angajat);
            
            updateCommand.CommandText = "update Angajat set nume = @nume, prenume= @prenume, locatie= @locatie, parola=@parola where id_angajat= @id_angajat";
            
            IDbDataParameter parameterNume = updateCommand.CreateParameter();
            parameterNume.ParameterName = "@nume";
            parameterNume.Value = angajat.Nume;
            updateCommand.Parameters.Add(parameterNume);
            IDbDataParameter parameterPrenume = updateCommand.CreateParameter();
            parameterPrenume.ParameterName = "@prenume";
            parameterPrenume.Value = angajat.Prenume;
            updateCommand.Parameters.Add(parameterPrenume);
            IDbDataParameter parameterLocatie = updateCommand.CreateParameter();
            parameterLocatie.ParameterName = "@locatie";
            parameterLocatie.Value = angajat.Locatie;
            updateCommand.Parameters.Add(parameterLocatie);
            IDbDataParameter parameterParola = updateCommand.CreateParameter();
            parameterParola.ParameterName = "@parola";
            parameterParola.Value = angajat.Parola;
            updateCommand.Parameters.Add(parameterParola);
            IDbDataParameter parameterIdAngajat = updateCommand.CreateParameter();
            parameterIdAngajat.ParameterName = "@id_angajat";
            parameterIdAngajat.Value = angajat.Id;
            updateCommand.Parameters.Add(parameterIdAngajat);
            
            int rows = updateCommand.ExecuteNonQuery();
            log.InfoFormat("Update {0} rows", rows);
            if (rows == 0)
                return angajat;
            return null;
        }
    }

    public Angajat? DeleteById(int id)
    {
        log.InfoFormat("Entering DeleteById with id: {0}",id);
        IDbConnection connection = DBUtils.getConnection(_props);
        using(var deleteCommand = connection.CreateCommand())
        {
            Angajat? angajat = FindById(id);
            
            deleteCommand.CommandText = "delete from Angajat where id_angajat= @id_angajat";
            
            IDbDataParameter parameterIdAngajat = deleteCommand.CreateParameter();
            parameterIdAngajat.ParameterName = "@id_angajat";
            parameterIdAngajat.Value = id;
            deleteCommand.Parameters.Add(parameterIdAngajat);
            
            int rows = deleteCommand.ExecuteNonQuery();
            log.InfoFormat("Delete {0} rows", rows);
            return angajat;
        }
    }

    public Angajat? FindById(int id)
    {
        log.InfoFormat("Entering FindById with id: {0}",id);
        IDbConnection connection = DBUtils.getConnection(_props);
        using(var selectCommand = connection.CreateCommand())
        {
            selectCommand.CommandText =
                "select * from Angajat where id_angajat = @id_angajat";
            IDbDataParameter parameterIdAngajat = selectCommand.CreateParameter();
            parameterIdAngajat.ParameterName = "@id_angajat"; 
            parameterIdAngajat.Value = id;
            selectCommand.Parameters.Add(parameterIdAngajat);

            using (var dataReader = selectCommand.ExecuteReader())
            {
                if(dataReader.Read())
                {
                    int idAngajat = dataReader.GetInt32(0);
                    string nume = dataReader.GetString(1);
                    string prenume = dataReader.GetString(2);
                    string locatie = dataReader.GetString(3);
                    string parola = dataReader.GetString(4);

                    Angajat angajat = new Angajat(nume, prenume, locatie, parola);
                    angajat.Id = idAngajat;
                    log.InfoFormat("Find angajat: {0}", angajat);
                    return angajat;
                }
            }
            log.Info("No angajat found");
            return null;
        }
    }

    public List<Angajat> FindAll()
    {
        log.Info("Entering FindAll");
        IDbConnection connection = DBUtils.getConnection(_props);
        using(var selectCommand = connection.CreateCommand())
        {
            selectCommand.CommandText =
                "select * from Angajat";
            List<Angajat> angajati = new List<Angajat>();
            using (var dataReader = selectCommand.ExecuteReader())
            {
                while(dataReader.Read())
                {
                    int idAngajat = dataReader.GetInt32(0);
                    string nume = dataReader.GetString(1);
                    string prenume = dataReader.GetString(2);
                    string locatie = dataReader.GetString(3);
                    string parola = dataReader.GetString(4);

                    Angajat angajat = new Angajat(nume, prenume, locatie, parola);
                    angajat.Id = idAngajat;

                    angajati.Add(angajat);
                }
            }
            log.InfoFormat("Find {0} angajati", angajati.Count);
            return angajati;
        }
    }

    public Angajat? findByFirstnameLastnamePassword(string firstname, string lastname, string password)
    {
        log.InfoFormat("Entering FindByFirstnameLastnamePassword with id: {0}",firstname + " " + lastname + " " + password);
        IDbConnection connection = DBUtils.getConnection(_props);
        using(var selectCommand = connection.CreateCommand())
        {
            selectCommand.CommandText =
                "select * from Angajat where prenume = @firstname and nume = @lastname and parola = @password";
            IDbDataParameter parameterFirstname = selectCommand.CreateParameter();
            parameterFirstname.ParameterName = "@firstname"; 
            parameterFirstname.Value = firstname;
            selectCommand.Parameters.Add(parameterFirstname);
            IDbDataParameter parameterLastname = selectCommand.CreateParameter();
            parameterLastname.ParameterName = "@lastname";
            parameterLastname.Value = lastname;
            selectCommand.Parameters.Add(parameterLastname);
            IDbDataParameter parameterPassword = selectCommand.CreateParameter();
            parameterPassword.ParameterName = "@password"; 
            parameterPassword.Value = password;
            selectCommand.Parameters.Add(parameterPassword);

            using (var dataReader = selectCommand.ExecuteReader())
            {
                if(dataReader.Read())
                {
                    int idAngajat = dataReader.GetInt32(0);
                    string nume = dataReader.GetString(1);
                    string prenume = dataReader.GetString(2);
                    string locatie = dataReader.GetString(3);
                    string parola = dataReader.GetString(4);

                    Angajat angajat = new Angajat(nume, prenume, locatie, parola);
                    angajat.Id = idAngajat;
                    log.InfoFormat("Find angajat: {0}", angajat);
                    return angajat;
                }
            }
            log.Info("No angajat found");
            return null;
        }
    }
}